tests/tt_metal/tt_metal/unit_tests_common/watcher/test_noc_sanitize.cpp

```diff
@@ -0,0 +1,125 @@
+
+    // Write to the input DRAM buffer
+    std::vector<uint32_t> input_vec = create_random_vector_of_bfloat16(
+        dram_buffer_size, 100, std::chrono::system_clock::now().time_since_epoch().count());
+    tt_metal::detail::WriteToBuffer(input_dram_buffer, input_vec);
+
+    // Write runtime args - update to a core that doesn't exist
+    output_dram_noc_xy.x = 16;
+    output_dram_noc_xy.y = 16;
+    tt_metal::SetRuntimeArgs(
+        program,
+        dram_copy_kernel,
+        core,
+        {l1_buffer_addr,
+        input_dram_buffer_addr,
+        (std::uint32_t)input_dram_noc_xy.x,
+        (std::uint32_t)input_dram_noc_xy.y,
+        output_dram_buffer_addr,
+        (std::uint32_t)output_dram_noc_xy.x,
+        (std::uint32_t)output_dram_noc_xy.y,
+        dram_buffer_size});
+
+    // Run the kernel, expect an exception here
+    try {
+        fixture->RunProgram(device, program);
+    } catch (std::runtime_error& e) {
+        const string expected = "Command Queue could not finish: device hang due to illegal NoC transaction. See build/watcher.log for details.";
+        const string error = string(e.what());
+        log_info(tt::LogTest, "Caught exception (one is expected in this test): {}", error);
+        EXPECT_TRUE(error.find(expected) != string::npos);
+    }
+
+    // We should be able to find the expected watcher error in the log as well.
+    CoreCoord phys_core = device->worker_core_from_logical_core(core);
+    string expected = "Device x, Core (x=x,y=x):    NAWW,*,*,*,*  brisc using noc0 tried to access core (16,16) L1[addr=0x00019020,len=102400]";
+    expected[7] = '0' + device->id();
+    expected[18] = '0' + phys_core.x;
+    expected[22] = '0' + phys_core.y;
+    EXPECT_TRUE(
+        FileContainsAllStrings(
+            fixture->log_file_name,
+            {expected}
+        )
+    );
+}
+
+// Run tests for host-side sanitization (uses functions that are from watcher.hpp).
+void CheckHostSanitization(Device *device) {
+    // Try reading from a core that doesn't exist
+    constexpr CoreCoord core = {16, 16};
+    uint64_t addr = 0;
+    uint32_t sz_bytes = 4;
+    try {
+        llrt::read_hex_vec_from_core(device->id(), core, addr, sz_bytes);
+    } catch (std::runtime_error& e) {
+        const string expected = "Host watcher: bad {} NOC coord {}\nread\n" + core.str();
+        const string error = string(e.what());
+        log_info(tt::LogTest, "Caught exception (one is expected in this test): {}", error);
+        EXPECT_TRUE(error.find(expected) != string::npos);
+    }
+}
+
+TEST_F(WatcherFixture, TestWatcherSanitize) {
+    // Skip this test for slow dipatch for now. Due to how llrt currently sits below device, it's
+    // tricky to check watcher server status from the finish loop for slow dispatch. Once issue #4363
+    // is resolved, we should add a check for print server handing in slow dispatch as well.
+    if (this->slow_dispatch_)
```

Comment:
Issue here is that fast dispatch is not supported yet for the remote device. This means that the test does not currently check the functionality on remote device. Plan is to leave this as-is for this PR, since we are actively bringing up fast dispatch for remote device.

Reply:
I have manually verified that this feature does work for the remote device